cmd/kava/cmd/rocksdb/compact.go

```diff
@@ -0,0 +1,194 @@
+//go:build rocksdb
+// +build rocksdb
+
+package rocksdb
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"os/signal"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"syscall"
+	"time"
+
+	"github.com/cosmos/cosmos-sdk/client"
+	"github.com/cosmos/cosmos-sdk/server"
+	"github.com/kava-labs/kava/cmd/kava/opendb"
+	"github.com/linxGnu/grocksdb"
+	"github.com/spf13/cobra"
+	"golang.org/x/exp/slices"
+
+	"github.com/tendermint/tendermint/libs/log"
+)
+
+var allowedDBs = []string{"application", "blockstore", "state"}
+
+var CompactRocksDBCmd = &cobra.Command{
+	Use: fmt.Sprintf(
+		"compact <%s>",
+		strings.Join(allowedDBs, "|"),
+	),
+	Short: "force compacts RocksDB",
+	Long: `This is a utility command that performs a force compaction on the state or
+blockstore. This should only be run once the node has stopped.`,
+	Args: cobra.ExactArgs(1),
+	RunE: func(cmd *cobra.Command, args []string) error {
+		logger := log.NewTMLogger(log.NewSyncWriter(os.Stdout))
+
+		clientCtx := client.GetClientContextFromCmd(cmd)
+		ctx := server.GetServerContextFromCmd(cmd)
+
+		if server.GetAppDBBackend(ctx.Viper) != "rocksdb" {
+			return errors.New("compaction is currently only supported with rocksdb")
+		}
+
+		if !slices.Contains(allowedDBs, args[0]) {
+			return fmt.Errorf(
+				"invalid db name, must be one of the following: %s",
+				strings.Join(allowedDBs, ", "),
+			)
+		}
+
+		compactRocksDBs(clientCtx.HomeDir, logger, args[0])
+		return nil
+	},
+}
+
+// compactRocksDBs performs a manual compaction on the given db.
+func compactRocksDBs(
+	rootDir string,
+	logger log.Logger,
+	dbName string,
+) error {
+	dbPath := filepath.Join(rootDir, "data", dbName+".db")
+
+	dbOpts, cfOpts, err := opendb.LoadLatestOptions(dbPath)
+	if err != nil {
+		return err
+	}
+
+	logger.Info("opening db", "path", dbPath)
+	db, _, err := grocksdb.OpenDbColumnFamilies(
+		dbOpts,
+		dbPath,
+		[]string{opendb.DefaultColumnFamilyName},
+		[]*grocksdb.Options{cfOpts},
+	)
+	if err != nil {
+		return err
+	}
+
+	if err != nil {
+		logger.Error("failed to initialize cometbft db", "path", dbPath, "err", err)
+		return fmt.Errorf("failed to open db %s %w", dbPath, err)
+	}
+	defer db.Close()
+
+	logColumnFamilyMetadata(db, logger)
+
+	logger.Info("starting compaction...", "db", dbPath)
+
+	done := make(chan bool, 1)
```

Comment:
Nice signal handling, however there are a couple subtle bugs.

Done may be sent but not selected by the compaction stats routine, or done may be sent after the ticker is selected but before GetProperty is called.  Since t`done` a buffered channel, the `done <- true` line will not block and the deferred close may run and leave enough time for GetProperty to called, still allowing use after close.

In addition, os.Exit() in the signal/interrupt handling will stop the program immediately and the defer in the command will not be run (https://pkg.go.dev/os#Exit).

The simplest fix is chaging `done` to a unbuffered channel and remove the os.Exit call and send on the done channel in the signal handler routine.  If I read the docs right, the call to disable manual compaction will cause CompactRange to return, resulting in done being called, stopping the stats routine and allowing a normal exit.

This separates concerns a bit - The signal handlers only job is to catch and cancel the blocking compaction.  The done channels job is to stop and wait for the stats routine to exit before continuing to the deferred close.

Note: If the stats routine needed cleanup a done channel wouldn't be enough since this doesn't guarantee the routine will return before exit, so if we needed that guarantee we'd have to use a context/cancel function with a wait group or some other pattern.